### ISSUE TITLE
When redirecting from AMP cache page to a page on our domain (cross-domain) the clientId changes. Only when coming in from google.co.uk (and not from google.com)

### ISSUE THREAD

**Issue by RonnieWinter** [ampproject/amphtml#19172]
## What's the issue?

Steps to recreate:
1. Go to google.co.uk and search for 'travelsupermarket new york holidays' and click on the amp result for travelsupermarket.
This will open https://www.google.co.uk/amp/s/www.travelsupermarket.com/en-gb/holidays/amp/united-states/new-york/

2. Go to the developer tools and click on the network tab, and filter by 'collect'. The 'cid' will look something like 'amp-lQHQiJPQJyDXxFQDUWOHrV34mX-1qABaNB80OXbbR9hj4UXwr3NS7OqVCB28tBKx'

3. Press on the 'search' button and this will redirect you to a url like this one: https://www.travelsupermarket.com/en-gb/holidays/results/5310a3fa823ab0522c0a32d4/528cc270e4b0ec1df53cbed9/2018-11-21/7/?room=A2
Go to the network tab and filter by 'collect'. Click on any of the google analytics requests, and check the 'cid' (clientId). This will be different to that of point 2 (which is incorrect).

Incidentally, if you repeat these steps but initiating the search from google.com (rather than google.co.uk) this problem does not occur and everything works as expected.

The meta tag `<meta name="amp-google-client-id-api" content="googleanalytics">` is present on the AMP page and the `ga('create', 'UA-XXXXX-Y', 'auto', {'useAmpClientId': true});` is present on the non AMP results page (the page we are redirected to)
You can read about this here: https://support.google.com/analytics/answer/7486764

**Comment by torch2424** [comment]
@lannka @calebcordry Is this a linker issue?

**Comment by ro2ni3** [comment]
> @ro2ni3 Is this a linker issue?
> 
> cc @lannka @calebcordry

I don't know if it is the AMP Linker. As far as I know, when initiating the journey from www.google.com everything works fine.

**Comment by calebcordry** [comment]
Hey @ro2ni3 thanks for reporting. This shouldn't be a linker issue (unless you are seeing this on safari only). I will take a look later today.

**Comment by ro2ni3** [comment]
> Hey @ro2ni3 thanks for reporting. This shouldn't be a linker issue (unless you are seeing this on safari only). I will take a look later today.

So far I have tested this on:
MacOs 10.13.6: Chrome (latest) and Safari (latest)
Ubuntu 18.04 LTS: Chrome (latest)
Windows 10: Chrome (latest)

I have noticed that this is a very intermittent issue. At one point during the day today, it started working fine, but now the bug is back.

**Comment by RonnieWinter** [comment]
Our data team has seen a lot of Safari iOS journeys that do not continue to the results page (coming in from the amp cached page) 
They are seeing lots of journeys that stop once they click on the Search button (at the point of going cross-domain). This therefore is more evidence that the users are being given a new ClientID.
Please have a look at one of the many journeys that appear to end at clicking the 'Search' button on the following screenshot:
![journey-stops-at-search](https://user-images.githubusercontent.com/32260852/48548741-a210a000-e8c5-11e8-8675-9a93df7e82b8.png)

**Comment by zhouyx** [comment]
@RonnieWinter To understand the issue better

1. You have opted in into the clientId API using metatag on AMP pages and with ga on non AMP pages
3. Only happens with Safari when page served from cache but not from google.com

I do know we have add support to subdomain via the `alternateUrl`  which should handle this specific use case. 

@rudygalfi @Gregable for more insight into the issue.

**Comment by zhouyx** [comment]
@RonnieWinter Talked with @Gregable offline. 

Looks like the problem here is different CID is stored under google.com and google.co.uk. 

Once user navigate from the cached AMP pages to your origin. The AMP clientId API tries to get the CID stored under google.com. If not found, based on the user ip, an alternateUrl will then be provided to attempt to get CID stored under the tld domain (here google.co.uk). 

In your case: User has visited both google.com and google.co.uk, where different clientId is stored under the two domains. So when they navigate from cached AMP pages under google.co.uk to your origin, the client id API respond with the incorrectly cid under google.com

We don't have a google solution to this. But you can use the AMP Linker to always pass the cid from cached AMP pages to your origin. 

Have a few questions:
1. Have you category the pageview data by Safari version? Since Linker is auto opted in in Safari 12, I don't expect a data loss with that Safari version. @avimehta please correct me if GA doesn't not auto read the cid from url. 
2. When did the pageview drop happen? Supposedly this should have been a long going issue.

**Comment by avimehta** [comment]
@zhouyx GA will read the value from linker. but this only works for IOS 12+ so all the other safari versions will still be broken.

**Comment by RonnieWinter** [comment]
> @RonnieWinter To understand the issue better
> 
>     1. You have opted in into the clientId API using metatag on AMP pages and with ga on non AMP pages
> 
>     2. Only happens with Safari when page served from cache but not from google.com
> 
> 
> I do know we have add support to subdomain via the `alternateUrl` which should handle this specific use case.
> 
> @rudygalfi @Gregable for more insight into the issue.

1. That is correct yes.
2. According to the data, yes Safari iOS seems to be hit the hardest by this bug. But when I recreate the bug, I am able to do so in Chrome, Firefox and Safari Desktop, Safari (emulating different iOS devices) and on different machines. 

There are 2 journeys affected by the ClientID not being carried over.

**1st Journey - from AMP cache page to non-AMP page**
- Go from the AMP cache page in https://www.google.co.uk/amp/s/www.travelsupermarket.com/en-gb/holidays/amp/spain/canary-islands/tenerife/ to a non-AMP page (by clicking the Search button)
for example: https://www.travelsupermarket.com/en-gb/holidays/results/5310a3fa823ab0522c0a32d4/54ad584b5218c8b2db215290/2018-11-30/7/?room=A2
The ClientID will change from the first to the second page

**2nd Journey - from non-AMP page to AMP cache page**
- Go to the home page: https://www.travelsupermarket.com/ You will be given a ClientID that looks like this "cid: 333161137.1542036408"
- Then go to the AMP cached page in https://www.google.co.uk/amp/s/www.travelsupermarket.com/en-gb/holidays/amp/spain/canary-islands/tenerife/ 
Your ClientID will have changed from "cid: 333161137.1542036408" to "cid: amp-gODr-a-txo60gXkvL7k9G5nb-XdXliCy0cZWwOaH4WmbBNC6_gnwqiDwoqBl-ygC"
This is incorrect, and in GA this appears as 2 different sessions.
- If we then proceed to click the search button, which takes us to the non-AMP page, the ClientID goes back to "cid: 333161137.1542036408"

**Differences between google.com and google.co.uk**
**Journey 1**
Now if you repeat the 1st Journey but go to the AMP cached page whose URL starts with google.com (https://www.google**.com**/amp/s/www.travelsupermarket.com/en-gb/holidays/amp/spain/canary-islands/tenerife/), then if we press the Search button which redirects us to a non-AMP page, the ClientID is carried over perfectly, and does not change.

**Journey 2**
Journey 2 bug happens regardless of whether you initiate your journey from google.com or google.co.uk

**Comment by RonnieWinter** [comment]
> @RonnieWinter Talked with @Gregable offline.
> 
> Looks like the problem here is different CID is stored under google.com and google.co.uk.
> 
> Once user navigate from the cached AMP pages to your origin. The AMP clientId API tries to get the CID stored under google.com. If not found, based on the user ip, an alternateUrl will then be provided to attempt to get CID stored under the tld domain (here google.co.uk).
> 
> In your case: User has visited both google.com and google.co.uk, where different clientId is stored under the two domains. So when they navigate from cached AMP pages under google.co.uk to your origin, the client id API respond with the incorrectly cid under google.com
> 
> We don't have a google solution to this. But you can use the AMP Linker to always pass the cid from cached AMP pages to your origin.
> 
> Have a few questions:
> 
>     1. Have you category the pageview data by Safari version? Since Linker is auto opted in in Safari 12, I don't expect a data loss with that Safari version. @avimehta please correct me if GA doesn't not auto read the cid from url.
> 
>     2. When did the pageview drop happen? Supposedly this should have been a long going issue.

In answer to 1 and 2: We have ruled out the problem being caused by Safari not sending pageviews on issue [19276](https://github.com/ampproject/amphtml/issues/19276) 
Even though Safari GA 'pageview' requests do not show up on the network tab, they are however hitting the GA endpoint. This is explained in issue 19276. I corroborated with our data guys in London that pageviews are indeed being tracked for Safari. So this is why issue 19276 has now been closed. 

Now, what we are actually seeing, is a significant drop in VTC that originates on Safari iOS devices. When the data guys looked closer, most journeys are stopping once the user jumps from an AMP cached page to a non-AMP page. They told me that this might not only be related to Safari iOS devices, as they can see other browsers and devices with interrupted journeys that stop once they go from the AMP cached page to the non AMP page.
As you can see from my previous reply where I explain 1st Journey and 2nd Journey, I have been able to recreate the broken journeys in all browsers on desktop (not only iOS Safari)

**Comment by RonnieWinter** [comment]
> @zhouyx yes, there is a support for `alternateUrl` for pages served from co.uk. Since I am located in US, it is harder to replicate the issue from here. @RonnieWinter Are you located in UK? if so, can you do the following and respond with the requests you see:
> 
>     1. Start with a clean browser (no existing cookies on google.com, amp domains and travelsupermarket domains)
> 
>     2. In the network tab of safari, click on "Preserve Requests"
> 
>     3. Filters requests by "cid"
> 
>     4. Go to Cache page
> 
>     5. Click on search to go to your domain.
> 
>     6. Reply with the requests that were sent/received.
> 
> 
> I am also interested in what the client id API responds with. If possible, please copy paste the json the is returned from the call to `https://ampcid.google.com/v1/publisher:getClientId` returns.

Here is the HAR output of running 1 to 6. (On 3. I filtered by 'collect'):
https://paste.ee/p/DisHo

Publisher:getClientId returns:
![screen shot 2018-11-16 at 11 46 17](https://user-images.githubusercontent.com/32260852/48619733-5e856700-e995-11e8-99c2-27867d888e90.png)

viewer:getClientId returns:
![screen shot 2018-11-16 at 11 46 09](https://user-images.githubusercontent.com/32260852/48619751-7230cd80-e995-11e8-8260-c178cb8b61a1.png)

**Comment by Gregable** [comment]
I can maybe provide a little help, but I'm not familiar with everything here, hopefully this is useful and not just noise.

> **1st Journey - from AMP cache page to non-AMP page**
> - Go from the AMP cache page in https://www.google.co.uk/amp/s/www.travelsupermarket.com/en-gb/holidays/amp/spain/canary-islands/tenerife/ to a non-AMP page (by clicking the Search button)
> for example: https://www.travelsupermarket.com/en-gb/holidays/results/5310a3fa823ab0522c0a32d4/54ad584b5218c8b2db215290/2018-11-30/7/?room=A2
> The ClientID will change from the first to the second page

I just tried this particular case using these exact URLs in Chrome at least. I used an incognito window so as to start with a clean set of cookies. In this case, I see consistent CID values on both pages.

What I looked for was in Chrome's network tab, I filtered for requests to www.google-analytics.com and looked at the URL parameters for the /collect requests.

There are a few possible reasons you'd see different results. 

1) The first is that you are looking at the wrong random-looking string for CID. 

In the screenshots you presented, I don't see the response body, only headers. The CID values are in the response body. The URL parameter strings that start with `AlzaSy...` are red herrings, they are constants that have nothing to do with the id itself.

Similarly, in the viewer response, you see a long `Set-Cookie: __Secure-AMPBCID=Anrau...`. The cookie value is not the CID and it is expected to change on every request. For clarity, but an unnecessary detail, is that this is an encrypted copy of your CID plus some added randomness, intended to constantly rotate.

If you want to see what the response body looks like, you can try something like:

```
$ curl https://ampcid.google.com/v1/viewer:getClientId -X POST      -d "{ 'publisherOrigin': 'https://amp.example.com',  'canonicalOrigin': 'https://example.com' }"      -H "Origin: https://www.google.com"      -H "Content-Type: application/json"
{
  "clientId": "G0vfQPQxEN4LGOKKrZ4k0Q"
}
```

Doing the same on the publisher:getClientId interface requires passing a cookie to it, so it is a bit harder to provide an example curl command. However, even if you did, it would be a different result. The clientId returned on viewer call is hashed with the publisher-side origin such as `www.travelsupermarket.com` to determine the final CID that is sent to analytics backends. On the `/viewer:getClientId` cache interface, this is done in javascript by the AMP viewer to reduce round trips. On your site in the `/publisher:getClientId` interface, this hashing is done by the ampcid.google.com server itself.

tl;dr: Debugging this by inspecting the requests to `ampcid.google.{com/TLD}` is a little non-intuitive.

2) The second is that you have two AMP cookies.

If your browser has an AMP cookie for `www.travelsupermarket.com` on both `ampcid.google.com` *and* `ampcid.google.co.uk`, then you can get into a bad state. 

What happens in a user journey from `google.co.uk` is what @zhouyx mentioned. The amp search result on `google.co.uk` will use the user's cookie on `google.co.uk` for generating a CID. Then when the user clicks to `www.travelsupermarket.com`, the analytics code there will make a request for a CID based on the `google.com` cookie. If present, it will just use that. If missing, it will try again by looking for a cookie on `google.co.uk`. If that's missing too, it will give up and mint a new local identifier.

For it to happen with a real user, that user must have viewed a `www.travelsupermarket.com` page in the AMP viewer on both `google.com` and `google.co.uk` in the past, thus producing unique cookies for both. This path may be rare among real users. It is showing up for you as your specific browser has tested paths on both google.com and google.co.uk. It's difficult to know the actual rate among real users of this corner case situation.

Unfortunately, there are very few options for improving this. The cookie spaces for `google.com` and `google.co.uk` are separate. The linker can help in some cases. In the long run, we hope that [signed exchanges](https://g.co/webpackagepreview) will resolve all of these issues.

**Comment by rudygalfi** [comment]
cc @jeffjose

**Comment by RonnieWinter** [comment]
> In the screenshots you presented, I don't see the response body, only headers. The CID values are in the response body. The URL parameter strings that start with `AlzaSy...` are red herrings, they are constants that have nothing to do with the id itself.

@Gregable Thanks very much for your reply. Sorry I forgot to include the body. 

The publisher:getClientId response body on the journey initiated from the amp cached page in google.co.uk is an empty body {}. I recorded my screen showing this (and you can see the empty response body at minute 1:24)
The video is [here](https://drive.google.com/file/d/1eDETavAJ0sLf3Klpy5mjhZCLKMyVUory/view?usp=sharing). The HAR export for the journey shown on this video is [here](https://paste.ee/p/kBC8D)

The publisher:getClientId response body on the journey initiated from the amp cached page in google.com does indeed contain the clientId and securityToken. I recorded my screen showing this (and you can see the body response at minute 1:08)
The video is [here](https://drive.google.com/file/d/1H7yXzZm4UE_jFHm7xCyjaZe2yi7ogdH1/view?usp=sharing). The HAR export for the journey shown on this video is [here](https://paste.ee/p/5kjD5)

**Comment by Gregable** [comment]
OK, so slightly different case than I originally understood.

As I understand it from watching your video, this would be the case where you start with no cookies, then go to `google.co.uk` which sets a CID-related cookie for `google.co.uk` and generates a related CID for Google-Analytics.

Next, you visit `www.travelsupermarket.com` and the Google-Analytics code makes a request to `google.com` for a CID. No cookie is set there, so this fails. The ip address of the request does not match one that we google.com would automatically redirect to `google.co.uk` and so no `alternateUrl` is returned. Since no `alternateUrl` is returned, no other attempt is made.

This can only happen if the user directly typed in `google.co.uk` and was not coming from an IP address that Google would have redirected to `google.co.uk`. Of course, that happens some amount of the time, but is also hopefully rare. How rare, I don't have data on.

**Comment by Gregable** [comment]
Ronnie, I think I may have poorly explained the scenario. I by no means suggest that UK based users are considered rare.

If a user has a UK based IP address, they will automatically be navigated to `google.co.uk` when making Google search queries. When the use selects an AMP document, a CID will be assigned associated with `google.co.uk`. If the user then visits `travelsupermarket.com`, the google analytics javascript will first query `google.com` for a CID. In this scenario, the user does not have a CID associated with `google.com`, so that request will fail. The javascript will follow that up with a 2nd request to `google.co.uk`, as determined by the user's IP address, to retrieve a CID associated with `google.co.uk`. Then it will use that CID.

This should all work correctly.

The corner case is when a single user's browser ends up clicking on amp documents in Google Search results on both `google.com` and `google.co.uk`. In this case, the user has a CID associated with both domains, and they differ. In this case, `travelsupermarket.com` will always get the CID that came from `google.com`, but the AMP Cache will get the CID from whichever domain the user is on at that time. 

Some scenarios where this can happen include:
 - a user with a non-UK ip address who directly visited `google.co.uk` and clickded on amp results and has also clicked on amp results on `google.com`.
 - a user who travels and thus has been on IP addresses associated with `google.com` and `google.co.uk` at different times.

I'm sure there are others.

**Comment by Tam-ara** [comment]
Hi evereybody, as the issue was cloesd: have the scenarios above been the only ones identified. We are experiencing similar problems, yet, aren't based in the UK nor have (a lot) of users based there.
Would be great to hear how you "solved" or worked around the issue. Thanks in advance!
